### PR TITLE
macOS: Integrate codesigning and enable hardened runtime

### PR DESCRIPTION
**Pull request by OatmealDome**
This PR integrates codesigning into the build system. It also enables the [hardened runtime](https://developer.apple.com/documentation/security/hardened_runtime) and adds the appropriate entitlements to DolphinQt. These are prerequisites to notarization.

To build:

```
cmake .. -DMACOS_CODE_SIGNING="ON" -DMACOS_CODE_SIGNING_IDENTITY="<identity common name or hash>"
```

Codesigning is disabled by default. In addition, the optional setting ``MACOS_CODE_SIGNING_IDENTITY_UPDATER`` can be set if the updater needs a different identity.

The entitlement `com.apple.security.automation.apple-events` is activated because I noticed `tccd` was raising an error about Dolphin attempting to use [Apple Events](https://developer.apple.com/documentation/applicationservices/apple_event_manager). I couldn't spot any obvious usage of Apple Events in the codebase, so I'm not sure why it's complaining in the first place. Either way, I added the entitlement so functionality is not unintentionally broken.

If this PR is merged, the next step would be to integrate notarization into the macOS buildbot if the project is willing to move forward with that. I can assist in any manner to get it up and running. I have already notarized some of my local Dolphin builds, and successfully tested them on macOS Catalina. Example binaries can be provided on request. (They have some personal information embedded in the app bundle, as I'm using my personal Apple Developer account, so I'd rather not have them be posted in a public forum.)

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt/Entitlements.plist [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+  <!-- Required for JIT usage -->
+  <key>com.apple.security.cs.allow-jit</key>
+  <true/>
+  <!-- Needed for GameCube microphone emulation -->
```
Comment (Codex-) [inline Source/Core/DolphinQt/Entitlements.plist:8]: This should probably be raised as a question in the PR body so it doesn't get missed too

Comment (OatmealDome) [reply Source/Core/DolphinQt/Entitlements.plist:8]: I assume you mean the Apple Events entitlement below? If so, then I've added a note in the PR body.

Comment (Codex-) [reply Source/Core/DolphinQt/Entitlements.plist:8]: Oof, yeah I did, mobile strikes again.

File: Source/Core/Common/MemoryUtil.cpp [hunk 2]
```diff
@@ -38,8 +38,12 @@ void* AllocateExecutableMemory(size_t size)
 #if defined(_WIN32)
   void* ptr = VirtualAlloc(nullptr, size, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
 #else
-  void* ptr =
-      mmap(nullptr, size, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_ANON | MAP_PRIVATE, -1, 0);
+  int flags = MAP_ANON | MAP_PRIVATE;
+#ifdef __APPLE__
+  flags |= MAP_JIT;
```
Comment (trevor403) [inline Source/Core/Common/MemoryUtil.cpp:43]: This has been available since Lion 10.7, I'm surprised that it is just being added now.

**Comment by ryanmcgrath** [comment]
Not sure if Dolphin still supports Sierra/High Sierra (10.12/10.13), but the README says it, so I figured I'd give a heads up: you may need to conditional the `MAP_JIT` for those platforms otherwise it'll crash - I had to work around this in Slippi like so:

``` c
static inline int determine_macos_jit_flag()
{
        static int jit_flag = -1;

        if (jit_flag == -1)
        {
                struct utsname name;
                uname(&name);

                // Kernel version 18 = Mojave
                jit_flag = (atoi(name.release) >= 18) ? MAP_JIT : 0;
        }

        return jit_flag;
}
```

``` c
flags |= determine_macos_jit_flag();
```

**Comment by OatmealDome** [comment]
@ryanmcgrath #9441 will probably supersede this PR, so I've forwarded your info there. Thanks for the heads up.

**Comment by OatmealDome** [comment]
Superseded by #9441!

**Pull request closed without merging**